noir-projects/aztec-nr/aztec/src/oracle/header.nr

```diff
@@ -30,22 +34,59 @@ pub fn get_header_at(block_number: u32, context: PrivateContext) -> Header {
-
-        // 4) Compute the block hash from the block header
-        let block_hash = header.hash();
+        // 3) Get the header hint of a given block from an oracle
+        let header_hint = get_header_at_internal(block_number);
 
-        // 5) Get the membership witness of the block in the archive
-        let witness = get_archive_membership_witness(last_archive_block_number, block_hash);
-
-        // 6) Check that the block is in the archive (i.e. the witness is valid)
-        assert(
-            context.historical_header.last_archive.root
-            == root_from_sibling_path(block_hash, witness.index, witness.path), "Proving membership of a block in archive failed"
+        // 4) We make sure that the header hint we received from the oracle exists in the state tree and is the actual header
+        // at the desired block number
+        constrain_get_header_at_internal(
+            header_hint,
+            block_number,
+            last_archive_block_number,
+            historical_header.last_archive.root
         );
 
-        // 7) Return the block header
-        header
+        // 8) Return the block header
+        header_hint
     }
 }
+
+fn constrain_get_header_at_internal(
+    header_hint: Header,
+    block_number: u32,
+    last_archive_block_number: u32,
+    last_archive_root: Field
+) {
+    // 1) Compute the block hash from the block header
+    let block_hash = header_hint.hash();
+
+    // 2) Get the membership witness of the block in the archive tree
+    let witness = get_archive_membership_witness(last_archive_block_number, block_hash);
+
+    // 3) Check that the block is in the archive (i.e. the witness is valid)
+    assert(
+        last_archive_root == root_from_sibling_path(block_hash, witness.index, witness.path), "Proving membership of a block in archive failed"
+    );
+
+    // 4) Check that the header hint has the same block number as the block number we are looking for, ensuring we are actually grabbing the header we specify
+    assert(
+        header_hint.global_variables.block_number as u32 == block_number, "Block number provided is not the same as the block number from the header hint"
+    );
+}
+
+#[test(should_fail_with = "Block number provided is not the same as the block number from the header hint")]
+fn fetching_a_valid_but_different_header_should_fail() {
+    let mut env = TestEnvironment::new();
+
+    // We get an arbitrary header at block 1
+    let old_header = get_header_at_internal(1);
+
+    // The next call to getHeader at block 5 should resolve with this old header.
+    // This header exists, but we make the call return it at an incorrect block number.
+    let _ = OracleMock::mock("getHeader").with_params((5)).returns(old_header.serialize()).times(1);
+
+    env.advance_block_to(6);
+
+    // We now try to get the header that our mock sets up. The oracle returns the old, incorrect header and should fail.
+    let _ = get_header_at(5, env.private());
```

Comment:
This is not wrong, but you could've just called `constrain_get_header_at_internal` in your test instead of mocking the oracle call. In more complex scenarios you may not be able to get away with this pattern.